### ISSUE TITLE
Use GCC diagnostic pragmas to tighten up code incrementally

### ISSUE THREAD

**Issue by mpe** [linuxppc/issues#228]
The recent pgd/pud bug (torvalds/linux@a58007621be3) highlighted that we're not building with `-Wconversion`.

The reason we're not is that it spews tens of thousands of warnings.

However we can enable/disable warnings for certain sections of code using GCC pragmas, eg:

```
diff --git a/arch/powerpc/include/asm/book3s/64/pgtable.h b/arch/powerpc/include/asm/book3s/64/pgtable.h
index d8c8d7c9df15..3114e3f368e2 100644
--- a/arch/powerpc/include/asm/book3s/64/pgtable.h
+++ b/arch/powerpc/include/asm/book3s/64/pgtable.h
@@ -904,7 +904,12 @@ static inline int pud_none(pud_t pud)
 
 static inline int pud_present(pud_t pud)
 {
+	__diag_push();
+	__diag_warn(GCC, 8, "-Wconversion", "ulong -> int");
+
 	return !!(pud_raw(pud) & cpu_to_be64(_PAGE_PRESENT));
+
+	__diag_pop();
 }
 
 extern struct page *pud_page(pud_t pud);
```

Would have caught the pgd/pud bug.

We don't want to add pragmas every where like above, but possibly we could enable it in certain files one-by-one.

The obvious first target would be `arch/powerpc/include/asm/book3s/64/pgtable.h`. Note we'd need to be careful to enable the warning only for powerpc code, not generic code pulled in via include. Otherwise we have to start fixing warnings in all the generic code.

**Comment by deece** [comment]
Maybe we should be fixing the warnings in generic code as we discover them, too

**Comment by mpe** [comment]
Yeah we could. That could become a big project though, eg. most of bitops.h needs work. And we may find people who think it's a waste of time and resist it, but who knows.
